<div class="alert alert-info">
<b>Комментарий студента:</b>

Алексей, привет. Давай на ты.

1. Прости - после твоих комментариев половину функций пришлось переделать, поэтому я загрузил новый файл...
2. Первая итерация проекта находится [здесь](https://github.com/mskhmelik/yandex_data_science/blob/master/Project_2/prj_bank_clients_churn_first_review.ipynb). Я надеюсь, этого будет достаточно, чтобы отследить изменения, но если есть какие-то проблемы - дай знать.

Пара вопросов:

1. Проект выглядел просто, когда я про него читал, но стал очень тяжелым, если делать его нормально, а не абы как. Я что-то не так делаю?
2. Кучу функций пришлось написать с нуля. Тот же `train_valid_test` без стратификации - это какой-то зашквар. Неужели нету каких-то библиотек? Я что-то упускаю?

</div>
    

# Отток клиентов банка

**Описание проекта**

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением F1-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте F1-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте AUC-ROC, сравнивайте её значение с F1-мерой.

План работы:

1. Загрузите и подготовьте данные.
2. Исследуйте баланс классов, обучите модель без учёта дисбаланса.
3. Улучшите качество модели, учитывая дисбаланс классов. Обучите разные модели и найдите лучшую.
4. Проведите финальное тестирование.

**Описание данных**

Признаки

1. `RowNumber` — индекс строки в данных
2. `CustomerId` — уникальный идентификатор клиента
3. `Surname` — фамилия
4. `CreditScore` — кредитный рейтинг
5. `Geography` — страна проживания
6. `Gender` — пол
7. `Age` — возраст
8. `Tenure` — сколько лет человек является клиентом банка
9. `Balance` — баланс на счёте
10. `NumOfProducts` — количество продуктов банка, используемых клиентом
11. `HasCrCard` — наличие кредитной карты
12. `IsActiveMember` — активность клиента
13. `EstimatedSalary` — предполагаемая зарплата

Целевой признак

1. `Exited` — факт ухода клиента

In [1]:
import pandas as pd
import optuna
import plotly.graph_objects as go

from caseconverter import snakecase
from ydata_profiling import ProfileReport
from IPython.display import display

from fast_ml import eda

from sklearn.model_selection import train_test_split
from sklearn.utils import resample

from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.dummy import DummyClassifier

from sklearn.metrics import (
    accuracy_score, f1_score, auc, roc_curve, roc_auc_score
)

In [2]:
FIG_WIDTH = 10 * 100
FIG_HEIGHT = 5 * 100
RANDOM_SEED = 42

In [3]:
try:
    raw_users = pd.read_csv('Churn.csv')
except:
    raw_users = pd.read_csv('/datasets/Churn.csv')

## Исследовательский анализ данных

Изучим основные зависимости в данных перед тем, как мы будем использовать их в алгоритмах машинного обучения.

Таблица-резюме:

In [4]:
display(eda.df_info(raw_users))

data_type data_type_grp num_unique_values  \
RowNumber           int64     Numerical             10000   
CustomerId          int64     Numerical             10000   
Surname            object   Categorical              2932   
CreditScore         int64     Numerical               460   
Geography          object   Categorical                 3   
Gender             object   Categorical                 2   
Age                 int64     Numerical                70   
Tenure            float64     Numerical                11   
Balance           float64     Numerical              6382   
NumOfProducts       int64     Numerical                 4   
HasCrCard           int64     Numerical                 2   
IsActiveMember      int64     Numerical                 2   
EstimatedSalary   float64     Numerical              9999   
Exited              int64     Numerical                 2   

                                              sample_unique_values  \
RowNumber                          [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]   
CustomerId       [15634602, 15647311, 15619304, 15701354, 15737...   
Surname          [Hargrave, Hill, Onio, Boni, Mitchell, Chu, Ba...   
CreditScore      [619, 608, 502, 699, 850, 645, 822, 376, 501, ...   
Geography                                 [France, Spain, Germany]   
Gender                                              [Female, Male]   
Age                       [42, 41, 39, 43, 44, 50, 29, 27, 31, 24]   
Tenure           [2.0, 1.0, 8.0, 7.0, 4.0, 6.0, 3.0, 10.0, 5.0,...   
Balance          [0.0, 83807.86, 159660.8, 125510.82, 113755.78...   
NumOfProducts                                         [1, 3, 2, 4]   
HasCrCard                                                   [1, 0]   
IsActiveMember                                              [1, 0]   
EstimatedSalary  [101348.88, 112542.58, 113931.57, 93826.63, 79...   
Exited                                                      [1, 0]   

                num_missing perc_missing  
RowNumber                 0          0.0  
CustomerId                0          0.0  
Surname                   0          0.0  
CreditScore               0          0.0  
Geography                 0          0.0  
Gender                    0          0.0  
Age                       0          0.0  
Tenure                  909         9.09  
Balance                   0          0.0  
NumOfProducts             0          0.0  
HasCrCard                 0          0.0  
IsActiveMember            0          0.0  
EstimatedSalary           0          0.0  
Exited                    0          0.0

Числовые распределения:

In [5]:
display(round(raw_users.describe().T, 2))

count         mean       std          min          25%  \
RowNumber        10000.0      5000.50   2886.90         1.00      2500.75   
CustomerId       10000.0  15690940.57  71936.19  15565701.00  15628528.25   
CreditScore      10000.0       650.53     96.65       350.00       584.00   
Age              10000.0        38.92     10.49        18.00        32.00   
Tenure            9091.0         5.00      2.89         0.00         2.00   
Balance          10000.0     76485.89  62397.41         0.00         0.00   
NumOfProducts    10000.0         1.53      0.58         1.00         1.00   
HasCrCard        10000.0         0.71      0.46         0.00         0.00   
IsActiveMember   10000.0         0.52      0.50         0.00         0.00   
EstimatedSalary  10000.0    100090.24  57510.49        11.58     51002.11   
Exited           10000.0         0.20      0.40         0.00         0.00   

                         50%          75%          max  
RowNumber            5000.50      7500.25     10000.00  
CustomerId       15690738.00  15753233.75  15815690.00  
CreditScore           652.00       718.00       850.00  
Age                    37.00        44.00        92.00  
Tenure                  5.00         7.00        10.00  
Balance             97198.54    127644.24    250898.09  
NumOfProducts           1.00         2.00         4.00  
HasCrCard               1.00         1.00         1.00  
IsActiveMember          1.00         1.00         1.00  
EstimatedSalary    100193.92    149388.25    199992.48  
Exited                  0.00         0.00         1.00

И детальный отчет:

In [6]:
ProfileReport(raw_users).to_widgets()

Render widgets:   0%|          | 0/1 [00:00<?, ?it/s]

Ключевые наблюдения из предварительного анализа набора данных:

1. **Демографические данные клиентов**: Набор данных содержит информацию о 10 000 клиентах с различными характеристиками. Средний возраст клиентов составляет 39 лет, с разбросом от 18 до 92 лет. В наборе данных представлены клиенты из трех стран: Франции, Испании и Германии. Кроме того, присутствует почти равномерное распределение мужчин и женщин.

2. **Финансовые показатели**: Набор данных включает финансовые метрики, такие как кредитные баллы, балансы счетов, количество продуктов и зарплаты. Средний кредитный балл составляет 651, что указывает на относительно хорошую кредитоспособность. Средний баланс счета составляет 76К долларов, с широким диапазоном значений. Клиенты имеют в среднем 1,53 продукта в банке, а зарплаты варьируются от 11 до 200К долларов.

3. **Отток и вовлеченность**: Около 20% клиентов в наборе данных покинули или прекратили использовать услуги банка. Приблизительно 52% клиентов являются активными участниками, что указывает на умеренный уровень вовлеченности. Средняя продолжительность работы с клиентом составляет 5 лет, хотя в этой колонке есть пропущенных значений для некоторых записей.

4. **Целевой признак**: У нас сть дисбаланс классов в целевом признаке `exited` - около 8К записей оставшихся пользователей и только 2К записей пользователей, которые перестали быть пользователями нашего банка. Это надо будет выровнять, чтобы избежать model bias.

# Подготовка данных для ML

Выполним эти преобразования на полном датасете:

1. Уберем колонки, которые не нужны для моделей: `RowNumber`, `CustomerId`, `Surname`.
2. Переведем колонку `gender` в бинарную: `1` - для мужчин и `0` - для женщин.
3. Названия колонок приведем `snake_case` регистру

И остальные после разделения на выборки, чтобы избежать data leakage:

1. Заполним пропуски в `tenure`.
2. Проведем `upsampling`.
3. Проведем стандартизацию численных признаков.
4. Проведем кодирование категориальных признаков. 

In [7]:
df_no_sampling = (
    raw_users
    .copy()
    .drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)
    .replace({'Male': 1, 'Female': 0})
    .rename(columns=lambda column: snakecase(column))
    .assign(geography=lambda df: df.geography.str.lower())
)

display(df_no_sampling.head())

credit_score geography  gender  age  tenure    balance  num_of_products  \
0           619    france       0   42     2.0       0.00                1   
1           608     spain       0   41     1.0   83807.86                1   
2           502    france       0   42     8.0  159660.80                3   
3           699    france       0   39     1.0       0.00                2   
4           850     spain       0   43     2.0  125510.82                1   

   has_cr_card  is_active_member  estimated_salary  exited  
0            1                 1         101348.88       1  
1            0                 1         112542.58       0  
2            1                 0         113931.57       1  
3            0                 0          93826.63       0  
4            1                 1          79084.10       0

Выглядит адекватно.

Дальше нам понадобится несколько функций для обработки данных. Первая - для разделенрия данных с использованием стратификации.

In [8]:
def split_data(df, target_column, test_size, valid_size, stratify_column):
    """
    Splits a DataFrame into training, validation and test sets.
    Args:
        - df (pd.DataFrame): DataFrame to split.
        - target_column (str): The name of the target column.
        - test_size (float): The proportion of the dataset to include in the test split.
        - valid_size (float): The proportion of the dataset to include in the validation split.
        - stratify_column (str): The name of the column to use for stratification.
    Returns:
        - list: A list containing the training, validation, and test feature and target DataFrames/Series.
    """
    df_train_valid, df_test, tgt_train_valid, tgt_test = train_test_split(
        df.drop(target_column, axis=1), df[target_column],
        test_size=test_size, stratify=df[stratify_column], random_state=RANDOM_SEED
    )
    df_train, df_valid, tgt_train, tgt_valid = train_test_split(
        df_train_valid, tgt_train_valid,
        test_size=(valid_size/(1-test_size)), stratify=tgt_train_valid, random_state=RANDOM_SEED
    )

    return [df_train, tgt_train, df_valid, tgt_valid, df_test, tgt_test]

Вторая - для `upsampling`.

In [9]:
def upsample_minority_class(ftr_train, tgt_train):
    """
    Upsamples the minority class in the training data.
    Args:
        - ftr_train (pd.DataFrame): DataFrame containing the training features.
        - tgt_train (pd.Series): Series containing the training targets.
    Returns:
        - pd.DataFrame: DataFrame containing the training features after downsampling.
        - pd.Series: Series containing the training targets after downsampling.
    """
    ftr_train_upsampled = pd.concat([
        ftr_train[tgt_train==0],
        resample(
            ftr_train[tgt_train==1],
            replace=True,
            n_samples=len(ftr_train[tgt_train==0]),
            random_state=RANDOM_SEED
        )
    ])

    tgt_train_upsampled = pd.concat([
        tgt_train[tgt_train==0],
        pd.Series([1]*len(ftr_train[tgt_train==0]), index=ftr_train[tgt_train==0].index)
    ])

    return ftr_train_upsampled, tgt_train_upsampled

Третья - для `downsampling`.

In [10]:
def downsample_majority_class(ftr_train, tgt_train):
    """
    Downsamples the majority class in the training data.
    Args:
        - ftr_train (pd.DataFrame): DataFrame containing the training features.
        - tgt_train (pd.Series): Series containing the training targets.
    Returns:
        - pd.DataFrame: DataFrame containing the training features after downsampling.
        - pd.Series: Series containing the training targets after downsampling.
    """
    index_to_remove = resample(
        ftr_train[tgt_train==0].index,
        replace=False,
        n_samples=len(ftr_train[tgt_train==0]) - len(ftr_train[tgt_train==1]),
        random_state=RANDOM_SEED
    )

    ftr_train_downsampled = ftr_train.drop(index_to_remove)
    tgt_train_downsampled = tgt_train.drop(index_to_remove)

    return ftr_train_downsampled, tgt_train_downsampled

Наконец, для предобработки данных.

In [11]:
def preprocess_and_split_data(df, label):
    """
    Preprocesses the data and splits it into training, validation and test sets.
    Args:
        - df (pd.DataFrame): DataFrame to preprocess and split.
        - label (str): Label for the dictionary to hold the split data.
    Returns:
        - dict: Dictionary containing preprocessed and split data.
    """
    splits = split_data(
        df=df, target_column='exited', 
        test_size=0.15, valid_size=0.15, stratify_column='exited', 
    )

    if label == 'upsampled':
        splits[0], splits[1] = upsample_minority_class(splits[0], splits[1])
    elif label == 'downsampled':
        splits[0], splits[1] = downsample_majority_class(splits[0], splits[1])

    data_splits = dict(zip(split_keys, splits))
    
    preprocessor.fit(data_splits['ftr_train'])

    for subset in split_keys:
        if 'ftr' in subset:
            data_transformed = preprocessor.transform(data_splits[subset])
            
            columns_transformed = num_cols + list(
                preprocessor.named_transformers_['cat']['encoder'].get_feature_names_out(cat_cols)
            )
            df_transformed = pd.DataFrame(
                data_transformed, columns=columns_transformed
            )
            df_transformed = pd.concat(
                [df_transformed, data_splits[subset][bin_cols].reset_index(drop=True)], axis=1
            )

            data_splits[subset] = df_transformed

    return data_splits

Разделим датасеты на выборки и выполним остальные преобразовния.

In [12]:
split_keys = ['ftr_train', 'tgt_train', 'ftr_valid', 'tgt_valid', 'ftr_test', 'tgt_test']
num_cols = ['credit_score', 'age', 'tenure', 'balance', 'num_of_products', 'estimated_salary']
cat_cols = ['geography']
bin_cols = ['has_cr_card', 'is_active_member', 'gender']

num_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler()),
])

cat_pipeline = Pipeline([
    ('encoder', OneHotEncoder(handle_unknown='ignore')),
])

preprocessor = ColumnTransformer([
    ('num', num_pipeline, num_cols),
    ('cat', cat_pipeline, cat_cols),
])

data_splits = {
    'no_sampling': preprocess_and_split_data(df_no_sampling, 'no_sampling'),
    'upsampled': preprocess_and_split_data(df_no_sampling, 'upsampled'),
    'downsampled': preprocess_and_split_data(df_no_sampling, 'downsampled')
}

Проверим, как сбалансировались классы.

In [13]:
def print_class_distribution(column, label):
    print(
        f"{'-' * 40}\n{label} distribution, #:\n{column.value_counts()}\n{'-' * 40}"
    )

print_class_distribution(data_splits['no_sampling']['tgt_train'], 'Exited without resampling')
print_class_distribution(data_splits['upsampled']['tgt_train'], 'Exited with upsampling')
print_class_distribution(data_splits['downsampled']['tgt_train'], 'Exited with downsampling')

----------------------------------------
Exited without resampling distribution, #:
0    5574
1    1426
Name: exited, dtype: int64
----------------------------------------
----------------------------------------
Exited with upsampling distribution, #:
0    5574
1    5574
dtype: int64
----------------------------------------
----------------------------------------
Exited with downsampling distribution, #:
0    1426
1    1426
Name: exited, dtype: int64
----------------------------------------


Теперь в `'upasmpled'` и `'downsampled'` датасетах классы сбалансированы.

Проверим, как выглядит один из датасетов.

In [14]:
data_splits['no_sampling']['ftr_valid'].head()

credit_score       age    tenure   balance  num_of_products  \
0      0.604610 -0.086317 -0.718147  0.872277        -0.906681   
1      0.894115 -0.086317  0.005948  1.634673        -0.906681   
2      0.490876 -1.614778  1.454139  0.369733        -0.906681   
3     -0.046776  0.200269 -1.442243  0.153721         2.531438   
4      1.028528  0.868971 -0.356100 -1.227786        -0.906681   

   estimated_salary  geography_france  geography_germany  geography_spain  \
0          0.962096               1.0                0.0              0.0   
1         -1.401904               0.0                0.0              1.0   
2         -0.547571               1.0                0.0              0.0   
3          1.556949               0.0                1.0              0.0   
4         -1.211556               1.0                0.0              0.0   

   has_cr_card  is_active_member  gender  
0            1                 1       1  
1            0                 1       1  
2            1                 0       1  
3            1                 0       0  
4            0                 1       0

В этом скрипте мы успешно обработали три набора данных, `no_sampling`, `upsampled` и `downsampled`. На первом этапе были определены переменные и настроен предпроцессор предварительной обработки для числовых и категориальных столбцов, бинарные столбцы остались без изменений.

Основной цикл скрипта выполняет три основные задачи для каждого набора данных:

1. Разделение данных: Используется функция `split_data`, чтобы разделить каждый набор данных на обучающие, валидационные и тестовые подмножества.

2. Подгонка препроцессора: Препроцессор обучается только на `train` данных. Это включает в себя изучение параметров для `imputation` наиболее часто встречающихся значений и стандартного масштабирования.

3. Трансформация данных: Используя подготовленный препроцессор, трансформируются признаки для каждого набора и типа колонок.

После выполнения этого скрипта у нас есть все подмножества данных, готовые для дальнейшего обучения модели и оценки.

# ML модели

## Функции для анализа

Перейдем к обучению моделей. Воспользуемся библиотекой `optuna` чтобы найти оптимальную модель.

Для начала - сделаем функцию, которая будет выбирать для нас лучшую модель.

In [15]:
def get_study(dataset: str, n_trials: int):
    """
    Trains and optimizes Logistic Regression, Decision Tree, and Random Forest models using Optuna.

    Args:
        - dataset (str): The label of the dataset to use from the 
          `data_splits` dictionary('no_sampling' or 'upsampled').
        - n_trials (int): The number of trials for Optuna optimization.

    Returns:
        - An Optuna study object, containing the optimal model and its parameters.
    """
    
    def objective(trial):
        classifier_name = trial.suggest_categorical(
            'classifier', ['LogisticRegression', 'DecisionTree', 'RandomForest']
        )
        
        # Suggest parameters based on the classifier
        if classifier_name == 'LogisticRegression':
            lr_c = trial.suggest_float('lr_c', 1e-6, 1e6, log=True)
            classifier_obj = LogisticRegression(
                C=lr_c, random_state=RANDOM_SEED
            )
        elif classifier_name == 'DecisionTree':
            dt_max_depth = trial.suggest_int('dt_max_depth', 1, 50)
            classifier_obj = DecisionTreeClassifier(
                max_depth=dt_max_depth, random_state=RANDOM_SEED
            )
        elif classifier_name == 'RandomForest':
            rf_max_depth = trial.suggest_int('rf_max_depth', 1, 100)
            rf_n_estimators = trial.suggest_int('rf_n_estimators', 100, 1000)
            classifier_obj = RandomForestClassifier(
                max_depth=rf_max_depth,
                n_estimators=rf_n_estimators,
                random_state=RANDOM_SEED
            )
            
        classifier_obj.fit(data_splits[dataset]['ftr_train'], data_splits[dataset]['tgt_train'])
        
        prediction = classifier_obj.predict(data_splits[dataset]['ftr_valid'])
        
        f1 = f1_score(data_splits[dataset]['tgt_valid'], prediction)
        
        return round(f1, 3)

    optuna.logging.set_verbosity(optuna.logging.WARNING)
    study = optuna.create_study(direction='maximize', sampler=optuna.samplers.TPESampler(seed=RANDOM_SEED))
    study.optimize(objective, n_trials=n_trials)
    
    return study

Теперь функцию, которая посчитает метрики для модели.

In [16]:
def evaluate_best_model(dataset: str, study: optuna.study.Study) -> dict[str, float]:
    """
    Given a dataset label (key in the data_splits dictionary) and an Optuna study object,
    this function retrieves the best model parameters from the study, instantiates a new
    model with these parameters, makes predictions on the validation subset,
    and calculates and returns the F1 score and ROC AUC score.

    Args:
        - dataset (str): Label of the dataset to use (should be a key in the data_splits dictionary).
        - study (optuna.study.Study): Optuna study object containing the results of model optimization.

    Returns:
        - dict: A dictionary with the F1 score and ROC AUC score for the validation subset.
        - sklearn Estimator: The best fitted model.
    """
    
    best_params = study.best_params

    if best_params['classifier'] == 'LogisticRegression':
        best_model = LogisticRegression(
            C=best_params['lr_c'], random_state=RANDOM_SEED
        )
    elif best_params['classifier'] == 'DecisionTree':
        best_model = DecisionTreeClassifier(
            max_depth=best_params['dt_max_depth'], random_state=RANDOM_SEED
        )
    elif best_params['classifier'] == 'RandomForest':
        best_model = RandomForestClassifier(
            max_depth=best_params['rf_max_depth'],
            n_estimators=best_params['rf_n_estimators'],
            random_state=RANDOM_SEED
        )

    best_model.fit(data_splits[dataset]['ftr_train'], data_splits[dataset]['tgt_train'])

    pred_valid = best_model.predict(data_splits[dataset]['ftr_valid'])
    pred_proba_valid = best_model.predict_proba(data_splits[dataset]['ftr_valid'])[:, 1]

    f1_valid = f1_score(data_splits[dataset]['tgt_valid'], pred_valid)
    roc_auc_valid = roc_auc_score(data_splits[dataset]['tgt_valid'], pred_proba_valid)

    return {
        'valid_f1': f1_valid,
        'valid_roc_auc': roc_auc_valid,
    }, best_model

Выведет метрики.

In [17]:
def print_best_params_and_metrics(best_params: dict, metrics: dict):
    """
    Prints the best parameters of the model and the performance metrics.

    Args:
        - best_params (dict): Dictionary of the best parameters from the model.
        - metrics (dict): Dictionary of performance metrics.

    Returns:
        - None
    """
    print('The best model parameters are:')
    for param, value in best_params.items():
        print(f"{param}: {value}")

    print('\nThe performance metrics of the model are:')
    for metric, value in metrics.items():
        print(f"{metric}: {value:.4f}")

Сделает сводный датафрейм.

In [18]:
def create_summary(studies: dict, evaluations: dict):
    """
    Create a summary dataframe from studies and evaluations dictionaries.

    Args:
        - studies (dict): Dictionary containing Optuna studies.
        - evaluations (dict): Dictionary containing evaluation results for each study.

    Returns:
        - pd.DataFrame: DataFrame summarizing the studies and their evaluation results.
    """

    data = {}

    for label in studies.keys():
        study_params = studies[label].best_params
        evaluation_metrics = evaluations[label][0]

        # Combine the study parameters and evaluation metrics into one dictionary
        data[label] = {**study_params, **evaluation_metrics}

    df_summary = pd.DataFrame(data).transpose()

    return df_summary

Нарисует график `ROC AUC`.

In [19]:
def plot_roc_auc(data_splits: dict, models: dict, dataset_type: str, title: str, label: str=None) -> go.Figure:
    """
    Plots ROC curves for the given datasets using Plotly.

    Args:
        - data_splits (dict): Dictionary where keys are types of data split ('train', 'valid', 'test')
          and values are tuples containing feature DataFrame and target Series for the given split.
        - models (dict): Dictionary where keys are model labels (matching keys in 'data_splits') 
          and values are trained model objects.
        - dataset_type (str): List of types of datasets to plot ('train', 'valid', 'test').
        - title (str): Title of the plot.
        - label (str, optional): Specific key from 'data_splits' and 'models' to use. If None, all keys are used.

    Returns:
        - fig (go.Figure): Plotly figure with ROC curves.
    """
    fig = go.Figure()

    labels = [label] if label else models.keys()

    for label in labels:
        best_model = models[label]
        
        ftr_data = data_splits[label]['ftr_' + dataset_type]
        tgt_data = data_splits[label]['tgt_' + dataset_type]

        probas = best_model.predict_proba(ftr_data)
        fpr, tpr, _ = roc_curve(tgt_data, probas[:, 1])
        roc_auc = auc(fpr, tpr)

        fig.add_trace(
            go.Scatter(
                x=fpr, y=tpr,
                mode='lines',
                name=f'{label} ROC (AUC = {roc_auc:.2f})'
            )
        )

    fig.add_trace(
        go.Scatter(
            x=[0, 1], y=[0, 1],
            mode='lines',
            name='No skill',
            line=dict(dash='dash', color='black')
        )
    )

    fig.update_layout(
        xaxis_title='False positive rate',
        yaxis_title='True positive rate',
        width=FIG_WIDTH, height=FIG_HEIGHT * 1.5,
        title=title,
        template='plotly_white'
    )

    return fig

И покажет результаты `optuna`.

In [20]:
def plot_study_results(label: str, studies: dict, evaluations: dict):
    """
    Plots the results for a given study.

    Args:
        - label (str): Label of the study to plot (should be a key in the studies and evaluations dictionaries).
        - studies (dict): Dictionary containing Optuna studies.
        - evaluations (dict): Dictionary containing valuation results for each study.

    Returns:
        - None
    """
    print_best_params_and_metrics(studies[label].best_params, evaluations[label][0])

    fig = optuna.visualization.plot_optimization_history(studies[label])
    fig.update_layout(template='plotly_white', width=FIG_WIDTH, height=FIG_HEIGHT)
    fig.show()

    fig = optuna.visualization.plot_slice(studies[label])
    fig.update_layout(template='plotly_white', width=FIG_WIDTH, height=FIG_HEIGHT)
    fig.update_xaxes(tickangle=-90)
    fig.show()

## Проверка метрик

В этой секции посмотрим на метрики, которые у нас получились для `no_sampling`, `downsampled` и `upsampled`.

### Резюме

Сведем результаты в отдельные переменные.

In [21]:
studies = {
    'no_sampling': get_study('no_sampling', n_trials=10),
    'downsampled': get_study('downsampled', n_trials=10),
    'upsampled': get_study('upsampled', n_trials=10),
}

evaluations = {
    'no_sampling': evaluate_best_model('no_sampling', studies['no_sampling']),
    'downsampled': evaluate_best_model('downsampled', studies['downsampled']),
    'upsampled': evaluate_best_model('upsampled', studies['upsampled']),
}

models = {
    'no_sampling': evaluations['no_sampling'][1],
    'downsampled': evaluations['downsampled'][1],
    'upsampled': evaluations['upsampled'][1],
}

Покажем результаты.

In [22]:
display(create_summary(studies, evaluations))

classifier rf_max_depth rf_n_estimators  valid_f1 valid_roc_auc
no_sampling  RandomForest           18             158  0.603175       0.86377
downsampled  RandomForest           18             158  0.607235      0.858571
upsampled    RandomForest           18             158  0.620321       0.85871

In [23]:
fig = plot_roc_auc(data_splits, models, 'valid', 'ROC curves based on validation datasets')
fig.show()

### Датасет без sampling

Возьмем датасет как есть и посмотрим на качество моделей.

In [24]:
plot_study_results('no_sampling', studies, evaluations)

The best model parameters are:
classifier: RandomForest
rf_max_depth: 18
rf_n_estimators: 158

The performance metrics of the model are:
valid_f1: 0.6032
valid_roc_auc: 0.8638


### Датасет с upsampling

Возьмем датасет c `upsampling` и посмотрим на качество моделей.

In [25]:
plot_study_results('upsampled', studies, evaluations)

The best model parameters are:
classifier: RandomForest
rf_max_depth: 18
rf_n_estimators: 158

The performance metrics of the model are:
valid_f1: 0.6203
valid_roc_auc: 0.8587


### Датасет с downsampling

Возьмем датасет c `downsampling` и посмотрим на качество моделей.

In [26]:
plot_study_results('downsampled', studies, evaluations)

The best model parameters are:
classifier: RandomForest
rf_max_depth: 18
rf_n_estimators: 158

The performance metrics of the model are:
valid_f1: 0.6072
valid_roc_auc: 0.8586


Результаты указывают на незначительные различия в производительности модели `RandomForest` в зависимости от того, были ли тренировочные данные `sampled`. Вот некоторые наблюдения:

1. **F1-оценка**: Метод `upsampling` показал наивысшую F1-оценку (0.620321) среди трех методов. Это говорит о том, что для этой конкретной модели и данных использование `upsampling` для балансировки классов в тренировочном наборе может улучшить баланс между `precision` и `recall`, которые измеряет F1-оценка.

2. **ROC AUC-оценка**: Однако модель, обученная на `downsampling`, имеет наивысшую ROC AUC-оценку (0.86377), что указывает на лучшую общую производительность с точки зрения как чувствительности (вероятность истинно положительного результата), так и специфичности (вероятность истинно отрицательного результата). Методы `downsampling` и `upsampling` имеют схожие ROC AUC-оценки (соответственно 0.858571 и 0.85871).

В общем, выбор метода может зависеть от того, какой показатель важнее для нашего конкретного приложения. Если вам больше важен баланс между точностью и полнотой, то возможно, стоит использовать `upsampling`. Но если более важна общее качество модели на основе ROC AUC-оценки, то возможно, будет лучше не использовать `sampling`.

Однако эти различия довольно малы, поэтому эффект от `sampling` не является значимым в данном случае. Более того, выбор гиперпараметров (максимальная глубина и количество оценщиков для случайного леса) остается стабильным во всех стратегиях выборки, что свидетельствует о устойвивости этих параметров независимо от `sampling`.

# Проверка на тест-выборке

Последним этапом, возьмем лучшую модель.

In [27]:
best_model = models['upsampled']

pred_test = best_model.predict(data_splits['upsampled']['ftr_test'])
pred_proba_test = best_model.predict_proba(data_splits['upsampled']['ftr_test'])[:, 1]

f1_test = f1_score(data_splits['upsampled']['tgt_test'], pred_test)
roc_auc_test = roc_auc_score(data_splits['upsampled']['tgt_test'], pred_proba_test)

И посмотрим на ее метрики.

In [28]:
print('Test f1 score:', round(f1_test, 4))
print('Test ROC AUC:', round(roc_auc_test, 4))
fig = plot_roc_auc(data_splits, models, 'test', 'ROC curves based on test datasets', 'upsampled')
fig.show()

Test f1 score: 0.5647
Test ROC AUC: 0.8406


# Выводы

На основе результатов экспериментов мы наблюдаем:

1. **Сравнение различных стратегий выборки:** Среди моделей, обученных с использованием различных стратегий выборки на тренировочных данных, модель, обученная с использованием метода `upsampling`, показала наивысший результат по метрике F1 (0.620321), немного превысив результаты, полученные `sampling` (0.603175) и при использовании `downsampling` (0.607235). Это говорит о том, что учет дисбаланса классов с помощью `upsampling` может положительно сказаться на способности модели совмещать точность и полноту.

2. **Площадь под кривой ROC (AUC-ROC):** Все три модели, независимо от `sampling`, показали схожие значения AUC-ROC, около 0.86, на валидационных данных. Это говорит о том, что модели получились сопоставимого качества с похожими границами определения целевого признака.

3. **Влияние `downsampling`:** Несмотря на то, что `downsampling` показывает немного лучший результат по F1-метрике, он все же ниже, чем у модели с `upsampling`. Это говорит о том, что, хотя `downsampling` может улучшить качество предсказаний по сравнению с датасетом as-is, оно может удалять потенциально полезную информацию, которую можно сохранить при `upsampling`.

4. **Производительность на тестовых данных:** Затем модель была оценена на тестовом наборе данных, где достигнут F1-счет 0.5647 и AUC-ROC 0.8406. Этот результат на невиданных данных немного ниже, чем результат на валидационном наборе. Снижение метрик может быть связано с переобучением модели на тренировочных данных, что является обычной проблемой в машинном обучении.

В итоге, увеличение выборки кажется наиболее эффективным методом борьбы с дисбалансом классов. Наконец, стоит рассмотреть другие метрики и деловые цели, помимо F1 и ROC-AUC, чтобы получить всестороннее понимание качества модели.